File: sdks/node/test/droid-policy.test.mjs
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import test from 'node:test';
import {
  assertMissing,
  configModuleUrl,
  createFixture,
} from '../test-support/droid-test-helpers.mjs';

test('Droid resolves hooksDisabled through official settings precedence', async (t) => {
  await t.test('user settings disable installation before runtime creation', async () => {
    const fixture = await createFixture({ settings: { hooksDisabled: true } });
    try {
      const result = await fixture.install();
      assert.equal(result.code, 1);
      assert.match(result.stderr, /hooksDisabled/);
      await assertMissing(path.join(fixture.agentDir, 'config.json'));
      await assertMissing(fixture.rootPath);
    } finally {
      await fixture.close();
    }
  });

  await t.test('user local settings override their base file', async () => {
    const fixture = await createFixture({
      settings: { hooksDisabled: true },
      localSettings: { hooksDisabled: false },
    });
    try {
      assert.equal((await fixture.install()).code, 0);
    } finally {
      await fixture.close();
    }
  });

  await t.test('legacy direct flags remain safety-compatible', async () => {
    const fixture = await createFixture({
      legacyConfig: { hooksDisabled: true, PreToolUse: [] },
      settings: { hooksDisabled: false },
    });
    try {
      const result = await fixture.install();
      assert.equal(result.code, 1);
      assert.match(result.stderr, new RegExp(fixture.legacyPath.replaceAll('\\', '\\\\'), 'i'));
    } finally {
      await fixture.close();
    }
  });

  await t.test('project policy takes precedence over user settings', async () => {
    const fixture = await createFixture({
      settings: { hooksDisabled: false },
      projectSettings: { hooksDisabled: true },
    });
    try {
      const result = await fixture.install();
      assert.equal(result.code, 1);
      assert.match(result.stderr, /project settings/);
      await assertMissing(path.join(fixture.agentDir, 'config.json'));
    } finally {
      await fixture.close();
    }
  });

  await t.test('project false locks out a lower user disable flag', async () => {
    const fixture = await createFixture({
      settings: { hooksDisabled: true },
      projectSettings: { hooksDisabled: false },
    });
    try {
      assert.equal((await fixture.install()).code, 0);
    } finally {
      await fixture.close();
    }
  });

  await t.test('project local settings override their base file', async () => {
    const fixture = await createFixture({
      projectSettings: { hooksDisabled: false },
      projectLocalSettings: { hooksDisabled: true },
    });
    try {
      const result = await fixture.install();
      assert.equal(result.code, 1);
      assert.match(result.stderr, /project local settings/);
    } finally {
      await fixture.close();
    }
  });
});

test('Droid treats allowManagedHooksOnly as an authoritative user-hook block', async () => {
  const { hookBlock } = await import(configModuleUrl);
  assert.deepEqual(hookBlock({
    policy: {
      allowManagedHooksOnlyBy: {
        filePath: '/managed/settings.json',
        label: 'Factory Droid system-managed settings',
      },
      preconditions: [],
    },
  }), {
    field: 'allowManagedHooksOnly',
    filePath: '/managed/settings.json',
    label: 'Factory Droid system-managed settings',
  });
});

test('Droid surfaces malformed read-only project policy before any write', async () => {
  const fixture = await createFixture({ projectSettings: '{ malformed' });
  try {
    const before = await readFile(fixture.projectSettingsPath, 'utf-8');
    const result = await fixture.install();
    assert.equal(result.code, 1);
    assert.match(result.stderr, /project settings/);
    assert.equal(await readFile(fixture.projectSettingsPath, 'utf-8'), before);
    await assertMissing(path.join(fixture.agentDir, 'config.json'));
    await assertMissing(fixture.rootPath);
  } finally {
    await fixture.close();
  }
});

File: sdks/go/cmd/elydora/plugins/droid_policy_test.go
package plugins

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestDroidUserSettingsDisableBeforeRuntimeCreation(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		settings: droidJSON(map[string]any{"hooksDisabled": true}),
	})
	err := fixture.plugin.Install(fixture.config)
	if err == nil || !strings.Contains(err.Error(), "hooksDisabled") {
		t.Fatalf("disabled hooks error = %v", err)
	}
	requireMissingDroidFile(t, fixture.runtimeConfig)
	requireMissingDroidFile(t, fixture.configPath)
}

func TestDroidUserLocalSettingsOverrideBaseSettings(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		settings:      droidJSON(map[string]any{"hooksDisabled": true}),
		localSettings: droidJSON(map[string]any{"hooksDisabled": false}),
	})
	installDroidFixture(t, fixture)
	if _, err := os.Stat(fixture.runtimeConfig); err != nil {
		t.Fatalf("runtime config is missing: %v", err)
	}
}

func TestDroidLegacyDirectFlagsRemainSafetyCompatible(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		legacy:   droidJSON(map[string]any{"hooksDisabled": true, "PreToolUse": []any{}}),
		settings: droidJSON(map[string]any{"hooksDisabled": false}),
	})
	err := fixture.plugin.Install(fixture.config)
	if err == nil || !strings.Contains(err.Error(), "legacy hooks") {
		t.Fatalf("legacy safety error = %v", err)
	}
	requireMissingDroidFile(t, fixture.runtimeConfig)
}

func TestDroidProjectPolicyHasExtensionOnlyPrecedence(t *testing.T) {
	t.Run("project blocks user allow", func(t *testing.T) {
		fixture := prepareDroidFixture(t, droidFixtureOptions{
			settings:        droidJSON(map[string]any{"hooksDisabled": false}),
			projectSettings: droidJSON(map[string]any{"hooksDisabled": true}),
		})
		err := fixture.plugin.Install(fixture.config)
		if err == nil || !strings.Contains(err.Error(), "project settings") {
			t.Fatalf("project policy error = %v", err)
		}
		requireMissingDroidFile(t, fixture.runtimeConfig)
	})
	t.Run("project allows through user block", func(t *testing.T) {
		fixture := prepareDroidFixture(t, droidFixtureOptions{
			settings:        droidJSON(map[string]any{"hooksDisabled": true}),
			projectSettings: droidJSON(map[string]any{"hooksDisabled": false}),
		})
		installDroidFixture(t, fixture)
	})
}

func TestDroidProjectLocalSettingsOverrideMatchingBase(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		projectSettings:      droidJSON(map[string]any{"hooksDisabled": false}),
		projectLocalSettings: droidJSON(map[string]any{"hooksDisabled": true}),
	})
	err := fixture.plugin.Install(fixture.config)
	if err == nil || !strings.Contains(err.Error(), "project local settings") {
		t.Fatalf("project local policy error = %v", err)
	}
}

func TestDroidProjectValuePrecedesDeeperFolderValue(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		projectSettings: droidJSON(map[string]any{"hooksDisabled": false}),
	})
	child := filepath.Join(fixture.workspaceDir, "packages", "console")
	if err := os.MkdirAll(child, 0755); err != nil {
		t.Fatalf("create child workspace: %v", err)
	}
	writeDroidTestObject(
		t,
		filepath.Join(child, ".factory", "settings.json"),
		map[string]any{"hooksDisabled": true},
	)
	if err := os.Chdir(child); err != nil {
		t.Fatalf("enter child workspace: %v", err)
	}
	installDroidFixture(t, fixture)
}

func TestDroidSystemManagedPolicyBlocksUserHooks(t *testing.T) {
	fixture := prepareDroidFixture(t, droidFixtureOptions{})
	writeDroidTestObject(
		t,
		fixture.systemSettingsPath,
		map[string]any{"allowManagedHooksOnly": true},
	)
	err := fixture.plugin.Install(fixture.config)
	if err == nil || !strings.Contains(err.Error(), "allowManagedHooksOnly") {
		t.Fatalf("managed policy error = %v", err)
	}
	requireMissingDroidFile(t, fixture.runtimeConfig)
}

func TestDroidMalformedReadOnlyProjectPolicyIsPreserved(t *testing.T) {
	malformed := "{ malformed"
	fixture := prepareDroidFixture(t, droidFixtureOptions{
		projectSettings: droidString(malformed),
	})
	err := fixture.plugin.Install(fixture.config)
	if err == nil || !strings.Contains(err.Error(), "project settings") {
		t.Fatalf("malformed project policy error = %v", err)
	}
	if readDroidTestFile(t, fixture.projectSettingsPath) != malformed {
		t.Fatal("malformed read-only policy changed")
	}
	requireMissingDroidFile(t, fixture.runtimeConfig)
	requireMissingDroidFile(t, fixture.configPath)
}
